rfd/0169-auto-updates-linux-agents.md

```diff
@@ -0,0 +1,267 @@
+
+```
+$ tree /var/lib/teleport
+/var/lib/teleport
+└── versions
+   ├── 15.0.0
+   │  ├── bin
+   │  │  ├── ...
+   │  │  ├── teleport-updater
+   │  │  └── teleport
+   │  └── etc
+   │     ├── ...
+   │     └── systemd
+   │        └── teleport.service
+   ├── 15.1.1
+   │  ├── bin
+   │  │  ├── ...
+   │  │  ├── teleport-updater
+   │  │  └── teleport
+   │  └── etc
+   │     ├── ...
+   │     └── systemd
+   │        └── teleport.service
+   └── updates.yaml
+$ ls -l /usr/local/bin/teleport
+/usr/local/bin/teleport -> /var/lib/teleport/versions/15.0.0/bin/teleport
+$ ls -l /usr/local/bin/teleport
+/usr/local/bin/teleport-updater -> /var/lib/teleport/versions/15.0.0/bin/teleport-updater
+$ ls -l /usr/local/lib/systemd/system/teleport.service
+/usr/local/lib/systemd/system/teleport.service -> /var/lib/teleport/versions/15.0.0/etc/systemd/teleport.service
+```
+
+updates.yaml:
+```
+version: v1
+proxy: mytenant.teleport.sh
+enabled: true
+active_version: 15.1.1
+```
+
+### Runtime
+
+The agent-updater will run as a periodically executing systemd service which runs every 10 minutes.
+The systemd service will run:
+```shell
+$ teleport-updater update
+```
+
+After it is installed, the `update` subcommand will no-op when executed until configured with the `teleport-updater` command:
+```shell
+$ teleport-updater enable --proxy mytenant.teleport.sh
+```
+
+If the proxy address is not provided with `--proxy`, the current proxy address from `teleport.yaml` is used.
+
+On servers without Teleport installed already, the `enable` subcommand will change the behavior of `teleport-update update` to update teleport and restart the existing agent, if running.
+It will also run update teleport immediately, to ensure that subsequent executions succeed.
+
+The `enable` subcommand will:
+1. Configure `updates.yaml` with the current proxy address and set `enabled` to true.
+2. Query the `/v1/webapi/ping` endpoint.
+3. If the current updater-managed version of Teleport is the latest, and teleport package is not installed, quit.
+4. If the current updater-managed version of Teleport is the latest, but the teleport package is installed, jump to (12).
+5. Download the desired Teleport tarball specified by `agent_version`.
+6. Verify the checksum.
+7. Extract the tarball to `/var/lib/teleport/versions/VERSION`.
+8. Replace any existing binaries or symlinks with symlinks to the current version.
+9. Restart the agent if the systemd service is already enabled.
+10. Set `active_version` in `updates.yaml` if successful or not enabled.
+11. Replace the old symlinks or binaries and quit (exit 1) if unsuccessful.
+12. Remove any `teleport` package if installed.
+13. Verify the symlinks to the active version still exists.
+14. Remove all stored versions of the agent except the current version and last working version.
+
+The `disable` subcommand will:
+1. Configure `updates.yaml` to set `enabled` to false.
+
+When `update` subcommand is otherwise executed, it will:
```

Comment:
This is covered above. It relies on systemd polling (every 10 min). Would you prefer a long-running daemon?

Reply:
If you imagined using a long-running watch connection, I would prefer to avoid creating an additional persistent connection to the proxy for every agent, given that additional load it would put on Cloud.

Also, it seems like that would require adding an unauthenticated watch endpoint? Is that what you imagined adding for `tctl autoupdate watch`? Seems like that would require careful rate limiting on the number of streams?

Reply:
I don't have a strong opinion, but I lean towards systemd oneshot for a few reasons:

- Simplifies the logic in the teleport-updater binary. Systemd makes it easy to configure retry logic, and it can set system statuses and notifications on each failure that can be monitored by other tools.
- Avoids running two Go binaries (agent and updater) continuously, reducing memory footprint.

If 10 minutes is too long, we can reduce the timer period. If necessary, new versions of the updater could configure the timer without requiring a package update.

That said, using systemd allows the user to configure the timer via overrides. This could benefit users who need to fine-tune their upgrade frequency / time window, but we may not want to permit that configurability, at least initially.

Reply:
I agree with Stephen, using a one-shot bin is easier. We don't have to deal with lifecycle, users can run it manually to debug, concurrency handling is easier, and it makes the self-updating approach saner (else the daemon has to reexec, then shoot itself in the head.

We can always convert later by changing the binary and its systemd unit.

Reply:
Consensus is one-shot for now